charts/spire/charts/spire-server/README.md

```diff
@@ -80,6 +84,10 @@ A Helm chart to install the SPIRE server.
 | nodeAttestor.k8sPsat.serviceAccountAllowList | list | `[]` |  |
 | nodeSelector | object | `{}` |  |
 | notifier.k8sbundle.namespace | string | `""` | Namespace to push the bundle into, if blank will default to SPIRE Server namespace |
+| persistence.accessMode | string | `"ReadWriteOnce"` |  |
+| persistence.enabled | bool | `true` | Enable persistence used by sqlite3 for spire-server, by the disk KeyStore, and/or by Tornjak. |
```

Comment:
And a few advantages. less volumes for admins to have to snapshot/backup.

Reply:
I agree with @kfox1111 on having 1 volume. Keeps things simple. If we need to have multiple volumes we can add in the future.

Reply:
Can we auto set the `persistence` if `sqllite3` is the db or `keymanager` is `disk`?

Reply:
Whats the value of disabling persistence?

Reply:
I've used it in several, related ways on other charts (there may be others)

1. In ci pipelines to speed up deployment of testing your own software, that uses spire. Provisioning the volume takes a bit to do, and being able to skip it reduces the overall runtime of the pipeline.  This enabled more rapid development/testing cycles.
2. In a ci pipeline where your deploying a test k8s. Similar to 1, not needing to provision a persistent volume driver at all can speed up deployment of the test cluster. Most valuable when dynamically allocating clusters in the pipeline
3. Some ci systems don't clean up after themselves very well (EX: cancellations). Its much easier to clean up helm charts that don't have pvc's.
4. Complex testing: while minikube has pv support in standalone node mode, it does not in its multinode mode. So if you want to do some nice multinode testing with it, just disabling persistence lets you deploy test instances of things such as spire, mysql, your code, etc in a nice scailable way, with stuff like antiaffinity, for quick testing of something that looks production like.

Reply:
Due the way these PVCs work any other workloads then spire-server will already define their own `persistentVolumeClaim`

See spire-server statefulset.

```yaml
  volumeClaimTemplates:
    {{- if eq (.Values.persistence.enabled | toString) "true" }}
    - metadata:
        name: spire-data
      spec:
        accessModes:
          - {{ .Values.persistence.accessMode | default "ReadWriteOnce" }}
        resources:
          requests:
            storage: {{ .Values.persistence.size }}
        {{- if .Values.persistence.storageClass }}
        storageClassName: {{ .Values.persistence.storageClass }}
        {{- end }}
    {{ end }}
```

If tornjak needs a PVC it will have to implement its own meaning you would also want to be able to choose its own storageClass and accessModes etc. That is just how it works. Going with this API simply means we will have to break the API as soon we add a PVC to tornjak. IMHO the persistence flag is to generalized, hence why it previously was at dataStorage specifically for the dataStorage volume.

maybe it should be 

```
dataStore:
  persistence:
    enabled: true
    …
```

So we can later do things like

```
tornjak:
  persistence:
    enabled: true
```

Oh and if we are about to disable persistence, then probably the statefulset should also be changed to a deployment if persistence is disabled. That is something I have seen in other charts as well. Happy to create a PR to patch those changes on top of this PR.

Reply:
> maybe it should be
> ```yaml
> dataStore:
>  persistence:
>    enabled: true
>    …
> ```
> So we can later do things like
> 
> ```yaml
> tornjak:
>  persistence:
>    enabled: true
> ```
and one for the disk keystore too. But, yeah, there'd be 3x the storage waste I think. I don't see the benefit in having so many volumes, especially once you have multiple replicas. But there is benefit in combining them for the end user.

> Oh and if we are about to disable persistence, then probably the statefulset should also be changed to a deployment if persistence is disabled. That is something I have seen in other charts as well. Happy to create a PR to patch those changes on top of this PR.

Other then for testing purposes, I don't think the user should ever be disabling persistence on the disk keystore. So at least 1 pvc is needed. Having the test case deployment look like the prod deployment in every way but deploy a little faster and not save state permanently is the ideal, so it to still be a statefulset is useful. Its totally ok for it to remain a statefulset and have no volumes. There are other properties of a statefulset that are useful outside pvs, such as persistent pod identity and lifecycle management that still make it a thing. I do volumeless statefulsets all the time.

> Imagine following usecase. I do want to run with an external database, that means I want to disable the persistence for sqlite, but I still want to run persistence for the keys and tornjak. How would I achieve that with this change?

The same way as today. You just don't turn off persistence. Nothing related to this has changed in this pr. Just the name of the key used to do the enabling/disabling as been renamed to better reflect the way its always worked, and to follow helm convention.

Reply:
Thought of one other thing, and I think its really important....

We can't add more volumes to the statefulset without a lot of upgrade pain on the users. Apparently volumeClaimTemplates is immutable. So if the user wanted to start out without, say, Tornjak, and then enable it later they would have to destroy their statefulset if it had sepaerate pvc's. 
```
The StatefulSet "test" is invalid: spec: Forbidden: updates to statefulset spec for fields other than
'replicas','template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and
'minReadySeconds' are forbidden
```

Reply:
We can still use a single pvc as long as they all store persistent data to `/run/spire/data` correct?  Thats what we are doing currently for sqllite3 and the disk keymanager. 

Im not familiar with Tornjak. If Tornjak is in same Pod then it would just need to do the same.

> Oh and if we are about to disable persistence, then probably the statefulset should also be changed to a deployment if persistence is disabled. That is something I have seen in other charts as well. Happy to create a PR to patch those changes on top of this PR.

I think this is a good idea to merge after this PR.

Reply:
> Thought of one other thing, and I think its really important....
> 
> We can't add more volumes to the statefulset without a lot of upgrade pain on the users. Apparently volumeClaimTemplates is immutable. So if the user wanted to start out without, say, Tornjak, and then enable it later they would have to destroy their statefulset if it had sepaerate pvc's.
> 
> ```
> The StatefulSet "test" is invalid: spec: Forbidden: updates to statefulset spec for fields other than
> 'replicas','template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and
> 'minReadySeconds' are forbidden
> ```

Fair point. Maybe we should switch to using a deployment with a separate PVC like they do in this chart. Probably for the reason you mentioning here? 
https://github.com/sigstore/helm-charts/blob/main/charts/rekor/templates/server/pvc.yaml

What do you think? Just trying to prevent API changes in followup releases.

Reply:
I think the discussion on StatefulSet vs Deployment is out of scope for this pr. I think its worth investigating separately as a follow on. For the changes in the pr, are we set on a single PV? If so then i think the changes here are good to go.